deps/lld/COFF/DriverUtils.cpp

```diff
@@ -638,10 +638,18 @@ void fixupExports() {
 
   if (config->killAt && config->machine == I386) {
     for (Export &e : config->exports) {
-      e.name = killAt(e.name, true);
-      e.exportName = killAt(e.exportName, false);
-      e.extName = killAt(e.extName, true);
-      e.symbolName = killAt(e.symbolName, true);
+      if (!e.name.empty() && e.name[0] == '?')
+        continue;
+      e.symbolName = e.name;
+      // Trim off the trailing decoration. Symbols will always have a
+      // starting prefix here (either _ for cdecl/stdcall, @ for fastcall
+      // or ? for C++ functions). Vectorcall functions won't have any
+      // fixed prefix, but the function base name will still be at least
+      // one char.
+      e.name = e.name.substr(0, e.name.find('@', 1));
+      // By making sure E.SymbolName != E.Name for decorated symbols,
+      // writeImportLibrary writes these symbols with the type
+      // IMPORT_NAME_UNDECORATE.
```

Comment:
We'll need to try to get this patch upstreamed. Many distributions insist on building zig against upstream LLD, which will break zig when it depends on the behavior in this patch. Plus it's just good open source citizenship.

I think you would be better able to make the case for the patch to LLD than me, since you have better understanding and you wrote the patch. However if you are not up for going through this process, let me know and I'll do it on your behalf.

Reply:
My idea is to eventually drop this patch and integrate the def->dll step functionality in the compiler itself (and possibly writing it in Zig). The whole thing is [really easy to implement](https://github.com/llvm-mirror/llvm/blob/master/lib/ToolDrivers/llvm-dlltool/DlltoolDriver.cpp) as the bulk of the work is done by `parseCOFFModuleDefinition` and `writeImportLibrary`.